File: performance/go/01/main.go
package main

import (
	"fmt"
	"sync"
	"time"
)

func sumRange(start, end int64, wg *sync.WaitGroup, result *int64) {
	defer wg.Done()
	var sum int64
	for i := start; i <= end; i++ {
		sum += i
	}
	*result = sum
}

func main() {
	start := time.Now()
	var wg sync.WaitGroup
	results := make([]int64, 4)
	partition := int64(250_000_000)

	for i := 0; i < 4; i++ {
		wg.Add(1)
		go sumRange(int64(i)*partition+1, (int64(i)+1)*partition, &wg, &results[i])
	}

	wg.Wait()

	var total int64
	for _, v := range results {
		total += v
	}

	duration := time.Since(start)
	fmt.Printf("Total: %d, Time: %v\n", total, duration)
}

File: performance/rust/01/main.rs
use std::thread;
use std::time::Instant;

fn sum_range(start: i64, end: i64) -> i64 {
    (start..=end).sum()
}

fn main() {
    let start = Instant::now();
    let partition = 250_000_000;
    let mut handles = vec![];

    for i in 0..4 {
        let s = i * partition + 1;
        let e = (i + 1) * partition;

        handles.push(thread::spawn(move || sum_range(s, e)));
    }

    let total: i64 = handles
        .into_iter()
        .map(|h| h.join().unwrap())
        .sum();

    let duration = start.elapsed();
    println!("Total: {}, Time: {:?}", total, duration);
}
